web_monitoring/diffing_server.py

```diff
@@ -325,35 +313,40 @@ def check_response_for_error(self, response):
 
 def _extract_encoding(headers, content):
     encoding = None
-    content_type = headers.get('Content-Type', '')
+    content_type = headers.get('Content-Type', '').lower()
     if 'charset=' in content_type:
         encoding = content_type.split('charset=')[-1]
-    if not encoding:
-        meta_tag_match = META_TAG_PATTERN.search(content, endpos=2048)
-        if meta_tag_match:
-            encoding = meta_tag_match.group(1).decode('ascii', errors='ignore')
-    if not encoding:
-        prolog_match = XML_PROLOG_PATTERN.search(content, endpos=2048)
-        if prolog_match:
-            encoding = prolog_match.group(1).decode('ascii', errors='ignore')
+        if encoding:
+            encoding = encoding.strip()
+    if not encoding and content:
+        # try to identify encoding using cchardet. Use up to 18kb of the
+        # content for detection. Its not necessary to use the full content
+        # as it could be huge. Also, if you use too little, detection is not
+        # accurate.
+        detected = cchardet.detect(content[:18432])
+        if detected:
+            detected_encoding = detected.get('encoding')
+            if detected_encoding:
+                encoding = detected_encoding.lower()
+
     # Handle common mistakes and errors in encoding names
     if encoding == 'iso-8559-1':
         encoding = 'iso-8859-1'
     # Windows-1252 is so commonly mislabeled, WHATWG recommends assuming it's a
     # mistake: https://encoding.spec.whatwg.org/#names-and-labels
     if encoding == 'iso-8859-1' and 'html' in content_type:
         encoding = 'windows-1252'
+    # Check if the selected encoding is known. If not, fallback to default.
+    try:
+        codecs.lookup(encoding)
+    except (LookupError, ValueError, TypeError):
+        encoding = 'ascii'
```

Comment:
This contradicts and prevents the first line of `_decode_body` from working; we default to UTF-8 encoding and only fall back to ASCII if that doesn’t work: https://github.com/edgi-govdata-archiving/web-monitoring-processing/blob/e602b3f94cdcbebc54db0861b38aa6f2a665b9bc/web_monitoring/diffing_server.py#L349-L350

Additionally, is there a case where we should expect a `ValueError` from `codecs.lookup`?

Reply:
This was a mistake, it should fallback to `utf-8`, I'll fix it.

Its possible to have a `ValueError` if we got a weird `encoding` value from headers or meta tag. I have seen all sorts of things like invalid unicode chars in there. Please lets keep it.

Reply:
Makes sense!